sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala

```diff
@@ -47,4 +48,44 @@ case class Md5(child: Expression)
       "org.apache.spark.unsafe.types.UTF8String.fromString" +
         s"(org.apache.commons.codec.digest.DigestUtils.md5Hex($c))")
   }
+
+}
+
+/**
+ * A function that returns a hash value of the argument
+ */
+case class Hash(children: Expression*) extends Expression {
+
+  override def foldable: Boolean = children.forall(_.foldable)
+  override def nullable: Boolean = children(0).nullable
+  override def dataType: DataType = IntegerType
+
+  override def eval(input: InternalRow): Any = {
+    val arglist = children.map(_.eval(input).asInstanceOf[AnyRef])
+    hashCode(arglist)
+  }
+
+  def hashCode(v: Any): Int = v match {
+    case null => 0
+    case arr: Array[_] => {
+      arr.foldLeft(0) { (acc, n) => acc + hashCode(n) }
+    }
+    case seq: Seq[_] => {
+      seq.foldLeft(0) { (acc, n) => acc + hashCode(n) }
+    }
+    case m: Map[_, _] => {
+      var res: Int = 0
+      m.foreach { case (k, v) => res += (hashCode(k) + hashCode(v)) }
+      res
+    }
+    case r: Row => {
+      var res: Int = 0
+      for (i <- 0 until r.length) {
+        res += 31 * res + hashCode(r.get(i))
+      }
+      res
+    }
+    case others => others.hashCode()
+  }
+
```

Comment:
Yes, so we need to reuse the wrapper, this will be helpful for the expression evaluation, since the result type always be the same, and even, we will not have box/unbox for primitive types.